File: monitors/producer-consumer.py
#! /usr/bin/env python

import threading
import time
import collections

TO_PRODUCE = 1000
BUFFER_SIZE = 10
PRODUCERS = 2
CONSUMERS = 2


class ProducerConsumer(object):
    def __init__(self, size):
        self.buffer = collections.deque([], size)
        self.mutex = threading.Lock()
        self.notFull = threading.Condition(self.mutex)
        self.notEmpty = threading.Condition(self.mutex)

    def append(self, data):
        with self.mutex:
            while len(self.buffer) == self.buffer.maxlen:
                self.notFull.wait()
            self.buffer.append(data)
            self.notEmpty.notify()

    def take(self):
        with self.mutex:
            while not self.buffer:
                self.notEmpty.wait()
            data = self.buffer.popleft()
            self.notFull.notify()
            return data


def producer(buffer):
    id = threading.current_thread().name
    print("Producer {}".format(id))

    for i in range(TO_PRODUCE):
        data = "{} i: {}".format(id, i)
        buffer.append(data)

def consumer(buffer):
    id = threading.current_thread().name
    print("Consumer {}".format(id))

    for i in range(TO_PRODUCE):
        data = buffer.take()
        print("{} read: {}".format(id, data))


def main():
    threads = []

    buffer = ProducerConsumer(BUFFER_SIZE)
    for i in range(CONSUMERS):
        c = threading.Thread(target=consumer, args=(buffer,))
        threads.append(c)

    for i in range(PRODUCERS):
        p = threading.Thread(target=producer, args=(buffer,))
        threads.append(p)

    # Start all threads
    for t in threads:
        t.start()

    # Wait for all threads to complete
    for t in threads:
        t.join()

    print("End")


if __name__ == "__main__":
    main()

File: monitors/ProducerConsumer.java
import java.util.Deque;
import java.util.LinkedList;


public class ProducerConsumer {
    static final int PRODUCERS = 2;
    static final int CONSUMERS = 2;


    public static void main(String[] args) {
        Thread[] threads = new Thread[PRODUCERS+CONSUMERS];
        int t = 0, i;
        PCMonitor monitor = new PCMonitor();


        System.out.printf("Counter");

        for (i = 0; i < CONSUMERS; i++) {
            threads[t] = new Thread(new Consumer(monitor));
            threads[t].start();
            t++;
        }

        for (i = 0; i < PRODUCERS; i++) {
            threads[t] = new Thread(new Producer(monitor));
            threads[t].start();
            t++;
        }

        for (i = 0; i < PRODUCERS+CONSUMERS; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {}
        }
    }

}

class Producer implements Runnable {
    static final int TO_PRODUCE = 1000;
    PCMonitor monitor;

    public Producer(PCMonitor monitor) {
        this.monitor = monitor;
    }

    @Override
    public void run() {
        long id = Thread.currentThread().getId();
        System.out.printf("Producer %d\n", id);

        for (int i = 0; i < TO_PRODUCE; i++ ) {
            monitor.append(i);
            System.out.printf("%d i: %d\n", id, i);
        }

        System.out.printf("Finished producer %d\n", id);
    }

}

class Consumer implements Runnable {
    static final int TO_CONSUME = 1000;

    PCMonitor monitor;
    public Consumer(PCMonitor monitor) {
        this.monitor = monitor;
    }

    @Override
    public void run() {
        long id = Thread.currentThread().getId();
        Integer data;

        System.out.printf("Consumer %d\n", id);

        for (int i = 0; i < TO_CONSUME; i++ ) {
            data = monitor.take();
            System.out.printf("%d read %d\n", id, data);
        }

        System.out.printf("Finished consumer %d\n", id);
    }

}

class PCMonitor {
    static final int BUFFER_SIZE = 10;

    int count = 0, in = 0, out = 0;
    Deque<Integer> buffer = new LinkedList<Integer>();

    synchronized public int take() {
        Integer data;

        while (buffer.isEmpty()) {
            try {
                wait();
            } catch (InterruptedException e) {}
        }

        data = buffer.remove();
        notifyAll();
        return data;
    }

    synchronized public void append(Integer data) {
        while (buffer.size() == BUFFER_SIZE) {
            try {
                wait();
            } catch (InterruptedException e) {}
        }
        buffer.add(data);
        notifyAll();
    }
}
